File: mlir/examples/dsp/SimpleBlocks/Output/TryDSPApps/BenchmarkTest/CCode/spaceCommunication.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define INPUT_LENGTH 100000000

// Function prototypes
double *getRangeOfVector(double start, int length, double increment);
void thresholdUp(const double *input, int length, double threshold,
                 char *output);
void space_modulate(const char *input, int *output, int length);
void transmit_and_receive(const int *signal, double *received_signal,
                          int length, double noise_level);
void demodulate(const double *signal, char *demodulated_data, int length);
void error_correction(const char *data, char *corrected);
void decode_data(const char *binary, char *decoded);

// Function implementations
void thresholdUp(const double *input, int length, double threshold,
                 char *output) {
  for (int i = 0; i < length; i++) {
    output[i] = (input[i] > threshold) ? '1' : '0';
  }
  output[length] = '\0';
}

void space_modulate(const char *input, int *output, int length) {
  for (int i = 0; i < length; i++) {
    output[i] = (input[i] == '1') ? 1 : -1;
  }
}

void transmit_and_receive(const int *signal, double *received_signal,
                          int length, double noise_level) {
  for (int i = 0; i < length; i++) {
    double noise = sin(signal[i]);
    received_signal[i] = signal[i] + noise;
  }
}

void demodulate(const double *signal, char *demodulated_data, int length) {
  for (int i = 0; i < length; i++) {
    demodulated_data[i] = (signal[i] > 0) ? '1' : '0';
  }
  demodulated_data[length] = '\0';
}

void error_correction(const char *data, char *corrected) {
  int length = strlen(data);
  int corrected_index = 0;
  for (int i = 0; i < length; i += 8) {
    int count = 0;
    for (int j = 0; j < 8; j++) {
      if (data[i + j] == '1')
        count++;
    }
    if (count % 2 == 0) {
      strncpy(&corrected[corrected_index], &data[i], 8);
    } else {
      corrected[corrected_index] = '0';
      strncpy(&corrected[corrected_index + 1], &data[i + 1], 7);
    }
    corrected_index += 8;
  }
  corrected[corrected_index] = '\0';
}

void decode_data(const char *binary, char *decoded) {
  int length = strlen(binary);
  int decoded_index = 0;
  for (int i = 0; i < length; i += 8) {
    char byte[9];
    strncpy(byte, &binary[i], 8);
    byte[8] = '\0';
    decoded[decoded_index++] = (char)strtol(byte, NULL, 2);
  }
  decoded[decoded_index] = '\0';
}

double *getRangeOfVector(double start, int length, double increment) {
  double *vector = malloc(length * sizeof(double));
  if (!vector) {
    perror("Memory allocation failed in getRangeOfVector");
    exit(EXIT_FAILURE);
  }
  for (int i = 0; i < length; i++) {
    vector[i] = start + i * increment;
  }
  return vector;
}

int main() {
  // Generate input vector
  double *input = getRangeOfVector(0, INPUT_LENGTH, 1);
  if (!input) {
    perror("Memory allocation failed for input");
    return EXIT_FAILURE;
  }

  // Threshold
  char *binary_sig = malloc(INPUT_LENGTH + 1);
  if (!binary_sig) {
    perror("Memory allocation failed for binary_sig");
    free(input);
    return EXIT_FAILURE;
  }
  thresholdUp(input, INPUT_LENGTH, 50, binary_sig);

  // Modulate
  int *modulated_signal = malloc(INPUT_LENGTH * sizeof(int));
  if (!modulated_signal) {
    perror("Memory allocation failed for modulated_signal");
    free(input);
    free(binary_sig);
    return EXIT_FAILURE;
  }
  space_modulate(binary_sig, modulated_signal, INPUT_LENGTH);

  // Transmit and receive (add noise)
  double *received_signal = malloc(INPUT_LENGTH * sizeof(double));
  if (!received_signal) {
    perror("Memory allocation failed for received_signal");
    free(input);
    free(binary_sig);
    free(modulated_signal);
    return EXIT_FAILURE;
  }
  transmit_and_receive(modulated_signal, received_signal, INPUT_LENGTH, 1.0);

  // Demodulate
  char *demodulated_data = malloc(INPUT_LENGTH + 1);
  if (!demodulated_data) {
    perror("Memory allocation failed for demodulated_data");
    free(input);
    free(binary_sig);
    free(modulated_signal);
    free(received_signal);
    return EXIT_FAILURE;
  }
  demodulate(received_signal, demodulated_data, INPUT_LENGTH);

  // Error correction
  char *corrected_data = malloc(INPUT_LENGTH + 1);
  if (!corrected_data) {
    perror("Memory allocation failed for corrected_data");
    free(input);
    free(binary_sig);
    free(modulated_signal);
    free(received_signal);
    free(demodulated_data);
    return EXIT_FAILURE;
  }
  error_correction(demodulated_data, corrected_data);

  // Decode data
  char *decoded_data = malloc((INPUT_LENGTH / 8) + 1);
  if (!decoded_data) {
    perror("Memory allocation failed for decoded_data");
    free(input);
    free(binary_sig);
    free(modulated_signal);
    free(received_signal);
    free(demodulated_data);
    free(corrected_data);
    return EXIT_FAILURE;
  }
  decode_data(corrected_data, decoded_data);

  printf("%c", corrected_data[8]);

  // Free allocated memory
  free(input);
  free(binary_sig);
  free(modulated_signal);
  free(received_signal);
  free(demodulated_data);
  free(corrected_data);
  free(decoded_data);

  return 0;
}

File: mlir/examples/dsp/SimpleBlocks/Output/TryDSPApps/BenchmarkTest/Matlab/spaceCommunication.m


function main()
    % Define constants
    INPUT_LENGTH = 100000000;
    
    % Generate input vector
    input = getRangeOfVector(0, INPUT_LENGTH, 1);
    
    % Threshold
    binary_sig = thresholdUp(input, INPUT_LENGTH, 50);
    
    % Modulate
    modulated_signal = space_modulate(binary_sig, INPUT_LENGTH);
    
    % Transmit and receive (add noise)
    received_signal = transmit_and_receive(modulated_signal, INPUT_LENGTH, 1.0);
    
    % Demodulate
    demodulated_data = demodulate(received_signal, INPUT_LENGTH);
    
    % Error correction
    corrected_data = error_correction(demodulated_data);
    
    % Decode data
    decoded_data = decode_data(corrected_data);
    
    % Display first corrected byte (equivalent to printing corrected_data[8] in C)
    fprintf('%c\n', corrected_data(9));
end

% Function to generate a vector with a given range and increment
function vector = getRangeOfVector(start, length, increment)
    vector = start:increment:(start + (length - 1) * increment);
end

% Thresholding function (creates a binary string from a vector)
function output = thresholdUp(input, length, threshold)
    output = char(zeros(1, length));  % Preallocate output
    output(input > threshold) = '1';
    output(input <= threshold) = '0';
end

% Space modulation: convert binary string to modulated signal
function output = space_modulate(input, length)
    output = zeros(1, length);
    output(input == '1') = 1;
    output(input == '0') = -1;
end

% Transmit and receive (add noise based on sine of the signal)
function received_signal = transmit_and_receive(signal, length, noise_level)
    received_signal = signal + sin(signal);  % Add noise (sine-based in this case)
end

% Demodulate: convert received signal back into binary data
function demodulated_data = demodulate(signal, length)
    demodulated_data = char(zeros(1, length));
    demodulated_data(signal > 0) = '1';
    demodulated_data(signal <= 0) = '0';
end

% Error correction function
function corrected = error_correction(data)
    length = numel(data);
    corrected = char(zeros(1, length));  % Preallocate corrected array
    corrected_index = 1;
    
    for i = 1:8:length
        segment = data(i:i+7);
        count = sum(segment == '1');
        
        if mod(count, 2) == 0
            corrected(corrected_index:corrected_index+7) = segment;
        else
            corrected(corrected_index) = '0';
            corrected(corrected_index+1:corrected_index+7) = segment(2:8);
        end
        
        corrected_index = corrected_index + 8;
    end
end

% Decode binary data to ASCII characters
function decoded = decode_data(binary)
    length = numel(binary);
    decoded = char(zeros(1, length / 8));  % Preallocate decoded data array
    decoded_index = 1;
    
    for i = 1:8:length
        byte = binary(i:i+7);
        decoded(decoded_index) = char(bin2dec(byte));
        decoded_index = decoded_index + 1;
    end
end
